Controller/Webhook/Transaction/RefundFailed.php

```diff
@@ -0,0 +1,147 @@
+use Magento\Sales\Model\OrderFactory;
+use Balancepay\Balancepay\Helper\Data;
+
+/**
+ * Balancepay transaction/refund-canceled webhook.
+ */
+class RefundFailed extends Action implements CsrfAwareActionInterface
+{
+    public const WEBHOOK_FAILED_NAME = 'transaction/refund-failed';
+
+    /**
+     * @var JsonFactory
+     */
+    private $jsonResultFactory;
+
+    /**
+     * @var BalancepayConfig
+     */
+    private $balancepayConfig;
+
+    /**
+     * @var RequestFactory
+     */
+    private $requestFactory;
+
+    /**
+     * @var Json
+     */
+    private $json;
+
+    /**
+     * @var OrderFactory
+     */
+    private $orderFactory;
+
+    /**
+     * @var Data
+     */
+    private $helperData;
+
+    /**
+     * @var WebhookRequestProcessor
+     */
+    private $webhookRequestProcessor;
+
+    /**
+     * Confirmed constructor.
+     *
+     * @param Context $context
+     * @param JsonFactory $jsonResultFactory
+     * @param BalancepayConfig $balancepayConfig
+     * @param RequestFactory $requestFactory
+     * @param Json $json
+     * @param OrderFactory $orderFactory
+     * @param Data $helperData
+     * @param WebhookRequestProcessor $webhookRequestProcessor
+     */
+    public function __construct(
+        Context $context,
+        JsonFactory $jsonResultFactory,
+        BalancepayConfig $balancepayConfig,
+        RequestFactory $requestFactory,
+        Json $json,
+        OrderFactory $orderFactory,
+        Data $helperData,
+        WebhookRequestProcessor $webhookRequestProcessor
+    ) {
+        parent::__construct($context);
+        $this->jsonResultFactory = $jsonResultFactory;
+        $this->balancepayConfig = $balancepayConfig;
+        $this->requestFactory = $requestFactory;
+        $this->json = $json;
+        $this->orderFactory = $orderFactory;
+        $this->helperData = $helperData;
+        $this->webhookRequestProcessor = $webhookRequestProcessor;
+    }
+
+    /**
+     * Execute
+     *
+     * @return ResponseInterface|ResultInterface
+     * @throws NoSuchEntityException
+     */
+    public function execute()
+    {
+        if (!$this->balancepayConfig->isActive()) {
+            return $this->resultFactory->create(ResultFactory::TYPE_FORWARD)->forward('noroute');
+        }
+        $content = $this->getRequest()->getContent();
+        $headers = $this->getRequest()->getHeaders()->toArray();
+
+        $this->balancepayConfig->log('Webhook\Checkout\Confirmed::execute() ', 'debug', [
+            'content' => $content,
+            'headers' => $headers,
+        ]);
+        return $this->webhookRequestProcessor->process($content, $headers, self::WEBHOOK_FAILED_NAME);
+    }
+
+    /**
+     * CreateCsrfValidationException
+     *
+     * @param RequestInterface $request
+     * @return InvalidRequestException|null
+     */
+    public function createCsrfValidationException(RequestInterface $request): ?InvalidRequestException
+    {
+        return null;
```

Comment:
same as before

Reply:
This starndard function and return value used for these CSRF process by magento.